src/services/item/plugins/published/plugins/search/meilisearch.ts

```diff
@@ -0,0 +1,397 @@
+        return index;
+      } catch (err) {
+        if (err instanceof MeiliSearchApiError && err.code === 'index_not_found') {
+          const task = await this.meilisearchClient.createIndex(name);
+          await this.meilisearchClient.waitForTask(task.taskUid);
+
+          // If main index just got created, fill it in the background
+          if (name === ACTIVE_INDEX) {
+            this.logger.info('Search index just created, rebuilding index...');
+            this.rebuildIndex();
+          }
+
+          const index = await this.meilisearchClient.getIndex(name);
+          this.indexDictionary[name] = index;
+          return index;
+        } else {
+          throw err;
+        }
+      }
+    }
+  }
+
+  // WORKS ONLY FOR PUBLISHED ITEMS
+  async search(queries: MultiSearchParams) {
+    const searchResult = await this.meilisearchClient.multiSearch(queries, {
+      hitsPerPage: 5,
+      attributesToHighlight: ['*'],
+    });
+
+    return searchResult;
+  }
+
+  private async parseItem(
+    item: Item,
+    categories: string[],
+    isPublishedRoot: boolean,
+    isHidden: boolean,
+  ): Promise<IndexItem> {
+    return {
+      id: item.id,
+      name: item.name,
+      description: this.removeHTMLTags(item.description),
+      type: item.type,
+      categories: categories,
+      content: await this.getContent(item),
+      isPublishedRoot: isPublishedRoot,
+      isHidden: isHidden,
+      createdAt: item.createdAt,
+      updatedAt: item.updatedAt,
+    };
+  }
+
+  // Retrieve searchable part inside an item
+  private async getContent(item: Item) {
+    switch (item.type) {
+      case ItemType.DOCUMENT:
+        return this.removeHTMLTags((item.extra as DocumentItemExtra).document.content); // better way to type extra safely?
+      case ItemType.LOCAL_FILE:
+        const localExtra = (item.extra as LocalFileItemExtra).file;
+        if (localExtra.mimetype === MimeTypes.PDF) {
+          return localExtra.content;
+        } else {
+          return '';
+        }
+      case ItemType.S3_FILE:
+        const s3extra = (item.extra as S3FileItemExtra).s3File;
+        if (s3extra.mimetype === MimeTypes.PDF) {
+          return s3extra.content;
+        } else {
+          return '';
+        }
+      default:
+        return '';
+    }
+  }
+
+  async indexOne(
+    item: Item,
+    repositories: Repositories,
+    targetIndex: ALLOWED_INDICES = ACTIVE_INDEX,
+  ): Promise<EnqueuedTask> {
+    return this.index([item], repositories, targetIndex);
+  }
+
+  async index(
+    items: Item[],
+    repositories: Repositories,
+    targetIndex: ALLOWED_INDICES = ACTIVE_INDEX,
+  ): Promise<EnqueuedTask> {
+    try {
+      // Get all descendants from the input items
+      const itemsToIndex: Item[] = [];
+      for (const item of items) {
+        itemsToIndex.push(item);
+        if (item.type === ItemType.FOLDER) {
+          itemsToIndex.push(...(await repositories.itemRepository.getDescendants(item)));
```

Comment:
Would it make sense to use `getManyDescendants`?

Reply:
good idea